pkg/services/pluginsintegration/clientmiddleware/cookies_middleware.go

```diff
@@ -0,0 +1,121 @@
+package clientmiddleware
+
+import (
+	"context"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/httpclient"
+	"github.com/grafana/grafana/pkg/components/simplejson"
+	"github.com/grafana/grafana/pkg/infra/httpclient/httpclientprovider"
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/services/contexthandler"
+	"github.com/grafana/grafana/pkg/services/datasources"
+	"github.com/grafana/grafana/pkg/util/proxyutil"
+)
+
+func NewCookiesMiddleware(skipCookiesNames []string) plugins.ClientMiddleware {
+	return plugins.ClientMiddlewareFunc(func(next plugins.Client) plugins.Client {
+		return &CookiesMiddleware{
+			next:             next,
+			skipCookiesNames: skipCookiesNames,
+		}
+	})
+}
+
+type CookiesMiddleware struct {
+	next             plugins.Client
+	skipCookiesNames []string
+}
+
+func (m *CookiesMiddleware) applyCookies(ctx context.Context, pCtx backend.PluginContext, req interface{}) (context.Context, error) {
+	reqCtx := contexthandler.FromContext(ctx)
+	// if request not for a datasource or no HTTP request context skip middleware
+	if req == nil || pCtx.DataSourceInstanceSettings == nil || reqCtx == nil || reqCtx.Req == nil {
+		return ctx, nil
+	}
+
+	settings := pCtx.DataSourceInstanceSettings
+	jsonDataBytes, err := simplejson.NewJson(settings.JSONData)
+	if err != nil {
+		return ctx, err
+	}
+
+	ds := &datasources.DataSource{
+		Id:       settings.ID,
+		OrgId:    pCtx.OrgID,
+		JsonData: jsonDataBytes,
+		Updated:  settings.Updated,
+	}
```

Comment:
JSON data bytes to struct in `DataSource`, back to bytes in `DataSourceInstanceSettings` and then back to `simplejson` is bad and feels like a hack, but can't figure out any other solution currently?